In [1]:
import numpy as np

In [2]:
import sklearn 

In [3]:
import pandas as pd

In [4]:
colnames =["duration","protocol_type","service","flag","src_bytes","dst_bytes", "land",
               "wrong_fragment","urgent","hot","num_failed_logins","logged_in", "num_compromised",
               "root_shell","su_attempted","num_root", "num_file_creations","num_shells", "num_access_files",
               "num_outbound_cmds","is_host_login","is_guest_login","count","srv_count","serror_rate",
              "srv_serror_rate","rerror_rate","srv_rerror_rate","same_srv_rate","diff_srv_rate",
               "srv_diff_host_rate","dst_host_count","dst_host_srv_count","dst_host_same_srv_rate",
               "dst_host_diff_srv_rate","dst_host_same_src_port_rate","dst_host_srv_diff_host_rate",
               "dst_host_serror_rate", "dst_host_srv_serror_rate","dst_host_rerror_rate",
               "dst_host_srv_rerror_rate","class"]

In [5]:
trainData = pd.read_csv("KDDTrain+_2.csv", header=None, names=colnames)

In [6]:
trainData.head()

duration protocol_type   service flag  src_bytes  dst_bytes  land  \
0         0           tcp  ftp_data   SF        491          0     0   
1         0           udp     other   SF        146          0     0   
2         0           tcp   private   S0          0          0     0   
3         0           tcp      http   SF        232       8153     0   
4         0           tcp      http   SF        199        420     0   

   wrong_fragment  urgent  hot  ...  dst_host_srv_count  \
0               0       0    0  ...                  25   
1               0       0    0  ...                   1   
2               0       0    0  ...                  26   
3               0       0    0  ...                 255   
4               0       0    0  ...                 255   

   dst_host_same_srv_rate  dst_host_diff_srv_rate  \
0                    0.17                    0.03   
1                    0.00                    0.60   
2                    0.10                    0.05   
3                    1.00                    0.00   
4                    1.00                    0.00   

   dst_host_same_src_port_rate  dst_host_srv_diff_host_rate  \
0                         0.17                         0.00   
1                         0.88                         0.00   
2                         0.00                         0.00   
3                         0.03                         0.04   
4                         0.00                         0.00   

   dst_host_serror_rate  dst_host_srv_serror_rate  dst_host_rerror_rate  \
0                  0.00                      0.00                  0.05   
1                  0.00                      0.00                  0.00   
2                  1.00                      1.00                  0.00   
3                  0.03                      0.01                  0.00   
4                  0.00                      0.00                  0.00   

   dst_host_srv_rerror_rate    class  
0                      0.00   normal  
1                      0.00   normal  
2                      0.00  neptune  
3                      0.01   normal  
4                      0.00   normal  

[5 rows x 42 columns]

In [7]:
trainData.describe()

duration     src_bytes     dst_bytes           land  \
count  125973.00000  1.259730e+05  1.259730e+05  125973.000000   
mean      287.14465  4.556674e+04  1.977911e+04       0.000198   
std      2604.51531  5.870331e+06  4.021269e+06       0.014086   
min         0.00000  0.000000e+00  0.000000e+00       0.000000   
25%         0.00000  0.000000e+00  0.000000e+00       0.000000   
50%         0.00000  4.400000e+01  0.000000e+00       0.000000   
75%         0.00000  2.760000e+02  5.160000e+02       0.000000   
max     42908.00000  1.379964e+09  1.309937e+09       1.000000   

       wrong_fragment         urgent            hot  num_failed_logins  \
count   125973.000000  125973.000000  125973.000000      125973.000000   
mean         0.022687       0.000111       0.204409           0.001222   
std          0.253530       0.014366       2.149968           0.045239   
min          0.000000       0.000000       0.000000           0.000000   
25%          0.000000       0.000000       0.000000           0.000000   
50%          0.000000       0.000000       0.000000           0.000000   
75%          0.000000       0.000000       0.000000           0.000000   
max          3.000000       3.000000      77.000000           5.000000   

           logged_in  num_compromised  ...  dst_host_count  \
count  125973.000000    125973.000000  ...   125973.000000   
mean        0.395736         0.279250  ...      182.148945   
std         0.489010        23.942042  ...       99.206213   
min         0.000000         0.000000  ...        0.000000   
25%         0.000000         0.000000  ...       82.000000   
50%         0.000000         0.000000  ...      255.000000   
75%         1.000000         0.000000  ...      255.000000   
max         1.000000      7479.000000  ...      255.000000   

       dst_host_srv_count  dst_host_same_srv_rate  dst_host_diff_srv_rate  \
count       125973.000000           125973.000000           125973.000000   
mean           115.653005                0.521242                0.082951   
std            110.702741                0.448949                0.188922   
min              0.000000                0.000000                0.000000   
25%             10.000000                0.050000                0.000000   
50%             63.000000                0.510000                0.020000   
75%            255.000000                1.000000                0.070000   
max            255.000000                1.000000                1.000000   

       dst_host_same_src_port_rate  dst_host_srv_diff_host_rate  \
count                125973.000000                125973.000000   
mean                      0.148379                     0.032542   
std                       0.308997                     0.112564   
min                       0.000000                     0.000000   
25%                       0.000000                     0.000000   
50%                       0.000000                     0.000000   
75%                       0.060000                     0.020000   
max                       1.000000                     1.000000   

       dst_host_serror_rate  dst_host_srv_serror_rate  dst_host_rerror_rate  \
count         125973.000000             125973.000000         125973.000000   
mean               0.284452                  0.278485              0.118832   
std                0.444784                  0.445669              0.306557   
min                0.000000                  0.000000              0.000000   
25%                0.000000                  0.000000              0.000000   
50%                0.000000                  0.000000              0.000000   
75%                1.000000                  1.000000              0.000000   
max                1.000000                  1.000000              1.000000   

       dst_host_srv_rerror_rate  
count             125973.000000  
mean                   0.120240  
std                    0.319459  
min                    0.000000  
25%       

In [9]:
#load the test dataset
testData = pd.read_csv("KDDTest+_2.csv", header=None, names=colnames)

In [10]:
testData.head()

duration protocol_type   service  flag  src_bytes  dst_bytes  land  \
0         0           tcp   private   REJ          0          0     0   
1         0           tcp   private   REJ          0          0     0   
2         2           tcp  ftp_data    SF      12983          0     0   
3         0          icmp     eco_i    SF         20          0     0   
4         1           tcp    telnet  RSTO          0         15     0   

   wrong_fragment  urgent  hot  ...  dst_host_srv_count  \
0               0       0    0  ...                  10   
1               0       0    0  ...                   1   
2               0       0    0  ...                  86   
3               0       0    0  ...                  57   
4               0       0    0  ...                  86   

   dst_host_same_srv_rate  dst_host_diff_srv_rate  \
0                    0.04                    0.06   
1                    0.00                    0.06   
2                    0.61                    0.04   
3                    1.00                    0.00   
4                    0.31                    0.17   

   dst_host_same_src_port_rate  dst_host_srv_diff_host_rate  \
0                         0.00                         0.00   
1                         0.00                         0.00   
2                         0.61                         0.02   
3                         1.00                         0.28   
4                         0.03                         0.02   

   dst_host_serror_rate  dst_host_srv_serror_rate  dst_host_rerror_rate  \
0                   0.0                       0.0                  1.00   
1                   0.0                       0.0                  1.00   
2                   0.0                       0.0                  0.00   
3                   0.0                       0.0                  0.00   
4                   0.0                       0.0                  0.83   

   dst_host_srv_rerror_rate    class  
0                      1.00  neptune  
1                      1.00  neptune  
2                      0.00   normal  
3                      0.00    saint  
4                      0.71    mscan  

[5 rows x 42 columns]

## Training and Test set Labels

In [12]:
trainData['class'].value_counts()

normal             67343
neptune            41214
satan               3633
ipsweep             3599
portsweep           2931
smurf               2646
nmap                1493
back                 956
teardrop             892
warezclient          890
pod                  201
guess_passwd          53
buffer_overflow       30
warezmaster           20
land                  18
imap                  11
rootkit               10
loadmodule             9
ftp_write              8
multihop               7
phf                    4
perl                   3
spy                    2
Name: class, dtype: int64

67343 is the count of normal traffic data, rest is all anomaly data with different types of attacks

In [13]:
testData['class'].value_counts()

normal             9711
neptune            4657
guess_passwd       1231
mscan               996
warezmaster         944
apache2             737
satan               735
processtable        685
smurf               665
back                359
snmpguess           331
saint               319
mailbomb            293
snmpgetattack       178
portsweep           157
ipsweep             141
httptunnel          133
nmap                 73
pod                  41
buffer_overflow      20
multihop             18
named                17
ps                   15
sendmail             14
rootkit              13
xterm                13
teardrop             12
xlock                 9
land                  7
xsnoop                4
ftp_write             3
sqlattack             2
loadmodule            2
phf                   2
udpstorm              2
worm                  2
perl                  2
imap                  1
Name: class, dtype: int64

# Step 1 : Data Preprocessing

Identify the features with categorical values. Convert them into numerical using Label/or One-Hot Encoding method. We can see there are three categorical columns, protocol_type, flag and service. We first need to find the unique categories within each categorical columns.

In [26]:
def CheckCategory(dataset):
    for c in colnames:
        if dataset[c].dtypes =='object':
         l= len(dataset[c].unique())
         print("Feature '{col}' has '{leng}' unique categories".format(col=c, leng=l))

In [27]:
CheckCategory(trainData)

Feature 'protocol_type' has '3' unique categories
Feature 'service' has '70' unique categories
Feature 'flag' has '11' unique categories
Feature 'class' has '23' unique categories


In [28]:
CheckCategory(testData)

Feature 'protocol_type' has '3' unique categories
Feature 'service' has '64' unique categories
Feature 'flag' has '11' unique categories
Feature 'class' has '38' unique categories


### Now lets print the categories

In [35]:
trainData['protocol_type'].value_counts().sort_values(ascending=False).head()

tcp     102689
udp      14993
icmp      8291
Name: protocol_type, dtype: int64

In [36]:
trainData['service'].value_counts().sort_values(ascending=False).head()

http        40338
private     21853
domain_u     9043
smtp         7313
ftp_data     6860
Name: service, dtype: int64

In [37]:
testData['flag'].unique()

array(['REJ', 'SF', 'RSTO', 'S0', 'RSTR', 'SH', 'S3', 'S2', 'S1',
       'RSTOS0', 'OTH'], dtype=object)

## One Hot Encoding

Now we would create dummy columns for each category in the categorical feature and use One-Hot coding to generate numerical values for them

In [38]:
category_column =['protocol_type', 'service', 'flag']

In [39]:
train_cat_values = trainData[category_column]

In [40]:
test_cat_values = testData[category_column]

In [41]:
train_cat_values.head()

protocol_type   service flag
0           tcp  ftp_data   SF
1           udp     other   SF
2           tcp   private   S0
3           tcp      http   SF
4           tcp      http   SF

Now create columns for each category

In [43]:
protocol_cat = sorted(train_cat_values['protocol_type'].unique())

In [45]:
# create columns corresponding to protocol_type
stringP= 'protocol_type_'
protocol_col = [stringP + x for x in protocol_cat]

In [46]:
protocol_col

['protocol_type_icmp', 'protocol_type_tcp', 'protocol_type_udp']

In [47]:
# create columns corresponding to categories in service 
service_cat =sorted(train_cat_values['service'].unique())
stringS='service_'
service_col =[stringS + x for x in service_cat]

In [48]:
# create columns corresponding to categories in flag
flag_cat =sorted(train_cat_values['flag'].unique())
stringF='flag_'
flag_col =[stringF + x for x in flag_cat]

In [49]:
train_category = protocol_col+service_col+flag_col

In [50]:
len(train_category)

84

The test data has 64 unique categories in service whereas train data has 70 unique categories in service. So we will have to create empty columns in test for 6 extra categories.

In [51]:
service_cat_test =sorted(test_cat_values['service'].unique())
stringS ='service_'
service_col_test =[stringS + x for x in service_cat_test]

In [52]:
test_category =protocol_col+service_col_test+flag_col

Now we would find the difference of service categories between train and test

In [53]:
difference = set(train_cat_values['service'])-set(test_cat_values['service'])

In [54]:
difference

{'aol', 'harvest', 'http_2784', 'http_8001', 'red_i', 'urh_i'}

The new dummy columns would look like below

In [55]:
print(train_category)

['protocol_type_icmp', 'protocol_type_tcp', 'protocol_type_udp', 'service_IRC', 'service_X11', 'service_Z39_50', 'service_aol', 'service_auth', 'service_bgp', 'service_courier', 'service_csnet_ns', 'service_ctf', 'service_daytime', 'service_discard', 'service_domain', 'service_domain_u', 'service_echo', 'service_eco_i', 'service_ecr_i', 'service_efs', 'service_exec', 'service_finger', 'service_ftp', 'service_ftp_data', 'service_gopher', 'service_harvest', 'service_hostnames', 'service_http', 'service_http_2784', 'service_http_443', 'service_http_8001', 'service_imap4', 'service_iso_tsap', 'service_klogin', 'service_kshell', 'service_ldap', 'service_link', 'service_login', 'service_mtp', 'service_name', 'service_netbios_dgm', 'service_netbios_ns', 'service_netbios_ssn', 'service_netstat', 'service_nnsp', 'service_nntp', 'service_ntp_u', 'service_other', 'service_pm_dump', 'service_pop_2', 'service_pop_3', 'service_printer', 'service_private', 'service_red_i', 'service_remote_job', 'serv

Now lets do One-Hot Encoding for these columns

In [58]:
from sklearn.preprocessing import OneHotEncoder

In [59]:
onehotenc = OneHotEncoder()

In [61]:
train_onehot= onehotenc.fit_transform(train_cat_values)

In [63]:
train_enc= pd.DataFrame(train_onehot.toarray(), columns=train_category)

In [64]:
train_enc.head()

protocol_type_icmp  protocol_type_tcp  protocol_type_udp  service_IRC  \
0                 0.0                1.0                0.0          0.0   
1                 0.0                0.0                1.0          0.0   
2                 0.0                1.0                0.0          0.0   
3                 0.0                1.0                0.0          0.0   
4                 0.0                1.0                0.0          0.0   

   service_X11  service_Z39_50  service_aol  service_auth  service_bgp  \
0          0.0             0.0          0.0           0.0          0.0   
1          0.0             0.0          0.0           0.0          0.0   
2          0.0             0.0          0.0           0.0          0.0   
3          0.0             0.0          0.0           0.0          0.0   
4          0.0             0.0          0.0           0.0          0.0   

   service_courier  ...  flag_REJ  flag_RSTO  flag_RSTOS0  flag_RSTR  flag_S0  \
0              0.0  ...       0.0        0.0          0.0        0.0      0.0   
1              0.0  ...       0.0        0.0          0.0        0.0      0.0   
2              0.0  ...       0.0        0.0          0.0        0.0      1.0   
3              0.0  ...       0.0        0.0          0.0        0.0      0.0   
4              0.0  ...       0.0        0.0          0.0        0.0      0.0   

   flag_S1  flag_S2  flag_S3  flag_SF  flag_SH  
0      0.0      0.0      0.0      1.0      0.0  
1      0.0      0.0      0.0      1.0      0.0  
2      0.0      0.0      0.0      0.0      0.0  
3      0.0      0.0      0.0      1.0      0.0  
4      0.0      0.0      0.0      1.0      0.0  

[5 rows x 84 columns]

In [65]:
test_onehot = onehotenc.fit_transform(test_cat_values)

In [67]:
test_enc = pd.DataFrame(test_onehot.toarray(), columns=test_category)

In [68]:
test_enc.head()

protocol_type_icmp  protocol_type_tcp  protocol_type_udp  service_IRC  \
0                 0.0                1.0                0.0          0.0   
1                 0.0                1.0                0.0          0.0   
2                 0.0                1.0                0.0          0.0   
3                 1.0                0.0                0.0          0.0   
4                 0.0                1.0                0.0          0.0   

   service_X11  service_Z39_50  service_auth  service_bgp  service_courier  \
0          0.0             0.0           0.0          0.0              0.0   
1          0.0             0.0           0.0          0.0              0.0   
2          0.0             0.0           0.0          0.0              0.0   
3          0.0             0.0           0.0          0.0              0.0   
4          0.0             0.0           0.0          0.0              0.0   

   service_csnet_ns  ...  flag_REJ  flag_RSTO  flag_RSTOS0  flag_RSTR  \
0               0.0  ...       1.0        0.0          0.0        0.0   
1               0.0  ...       1.0        0.0          0.0        0.0   
2               0.0  ...       0.0        0.0          0.0        0.0   
3               0.0  ...       0.0        0.0          0.0        0.0   
4               0.0  ...       0.0        1.0          0.0        0.0   

   flag_S0  flag_S1  flag_S2  flag_S3  flag_SF  flag_SH  
0      0.0      0.0      0.0      0.0      0.0      0.0  
1      0.0      0.0      0.0      0.0      0.0      0.0  
2      0.0      0.0      0.0      0.0      1.0      0.0  
3      0.0      0.0      0.0      0.0      1.0      0.0  
4      0.0      0.0      0.0      0.0      0.0      0.0  

[5 rows x 78 columns]

Now add 6 missing service category columns to testdata and assign them 0 value

In [70]:
difference = ['service_' + x for x in difference]

In [71]:
difference

['service_http_8001',
 'service_http_2784',
 'service_red_i',
 'service_harvest',
 'service_urh_i',
 'service_aol']

In [72]:
for col in difference:
    test_enc[col]=0

In [73]:
test_enc.head()

protocol_type_icmp  protocol_type_tcp  protocol_type_udp  service_IRC  \
0                 0.0                1.0                0.0          0.0   
1                 0.0                1.0                0.0          0.0   
2                 0.0                1.0                0.0          0.0   
3                 1.0                0.0                0.0          0.0   
4                 0.0                1.0                0.0          0.0   

   service_X11  service_Z39_50  service_auth  service_bgp  service_courier  \
0          0.0             0.0           0.0          0.0              0.0   
1          0.0             0.0           0.0          0.0              0.0   
2          0.0             0.0           0.0          0.0              0.0   
3          0.0             0.0           0.0          0.0              0.0   
4          0.0             0.0           0.0          0.0              0.0   

   service_csnet_ns  ...  flag_S2  flag_S3  flag_SF  flag_SH  \
0               0.0  ...      0.0      0.0      0.0      0.0   
1               0.0  ...      0.0      0.0      0.0      0.0   
2               0.0  ...      0.0      0.0      1.0      0.0   
3               0.0  ...      0.0      0.0      1.0      0.0   
4               0.0  ...      0.0      0.0      0.0      0.0   

   service_http_8001  service_http_2784  service_red_i  service_harvest  \
0                  0                  0              0                0   
1                  0                  0              0                0   
2                  0                  0              0                0   
3                  0                  0              0                0   
4                  0                  0              0                0   

   service_urh_i  service_aol  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 84 columns]

Both the train and test category have 84 columns. Now lets drop these categorical columns from main dataset and add the onehot encoded columns

In [74]:
train_new = trainData.join(train_enc)

In [75]:
test_new = testData.join(test_enc)

In [76]:
train_new.drop('flag',axis=1, inplace=True)

In [77]:
train_new.drop('service',axis=1, inplace=True)

In [78]:
train_new.drop('protocol_type',axis=1, inplace=True)

In [80]:
train_new.shape

(125973, 123)

In [81]:
#now drop the columns from the test dataset
test_new.drop('protocol_type',axis=1, inplace=True)

In [82]:
test_new.drop('flag',axis=1, inplace=True)
test_new.drop('service',axis=1, inplace=True)

In [83]:
test_new.shape

(22544, 123)

In [84]:
test_new.head()

duration  src_bytes  dst_bytes  land  wrong_fragment  urgent  hot  \
0         0          0          0     0               0       0    0   
1         0          0          0     0               0       0    0   
2         2      12983          0     0               0       0    0   
3         0         20          0     0               0       0    0   
4         1          0         15     0               0       0    0   

   num_failed_logins  logged_in  num_compromised  ...  flag_S2  flag_S3  \
0                  0          0                0  ...      0.0      0.0   
1                  0          0                0  ...      0.0      0.0   
2                  0          0                0  ...      0.0      0.0   
3                  0          0                0  ...      0.0      0.0   
4                  0          0                0  ...      0.0      0.0   

   flag_SF  flag_SH  service_http_8001  service_http_2784  service_red_i  \
0      0.0      0.0                  0                  0              0   
1      0.0      0.0                  0                  0              0   
2      1.0      0.0                  0                  0              0   
3      1.0      0.0                  0                  0              0   
4      0.0      0.0                  0                  0              0   

   service_harvest  service_urh_i  service_aol  
0                0              0            0  
1                0              0            0  
2                0              0            0  
3                0              0            0  
4                0              0            0  

[5 rows x 123 columns]

# Split the dataset as per attack type

The class labels in our dataset have different type of attacks, we need to identify the type of attacks and group them. Assign a number to each group. There are 4 types of attacks, namely, Denial of Service (DoS), Probing (Probe), Remote to User (R2L), User to Root (U2R). So for example, if the labels are given as 'normal', 'neptune', 'pod', 'smurf', 'portsweep', 'satan', 'guess_passwd', 'imap', 'perl', 'loadmodule', etc. We first group them as the type of attack, like we can see 'neptune, 'pod', 'smurf' are DoS attacks; 'portsweep' and 'satan' are Probe attacks ; 'guess_passwd', 'imap' are R2L attacks and 'perl', 'loadmodule' are U2R attacks. Now we can assign numbers normal =0 , DoS =1, Probe=2, R2L=3, U2R=4 and create four different datasets. 

### Rename attack class label; normal=0 , DoS =1, Probe=2, R2L=3, U2R=4 

In [86]:
# class column
label_train=train_new['class']
label_test=test_new['class']

In [87]:
# identify the type of attack and change the attack labels to numbers
label_train_new = label_train.replace({ 'normal' : 0, 'neptune' : 1 ,'back': 1, 'land': 1, 'pod': 1, 'smurf': 1, 'teardrop': 1,'mailbomb': 1, 'apache2': 1, 'processtable': 1, 'udpstorm': 1, 'worm': 1,
                           'ipsweep' : 2,'nmap' : 2,'portsweep' : 2,'satan' : 2,'mscan' : 2,'saint' : 2
                           ,'ftp_write': 3,'guess_passwd': 3,'imap': 3,'multihop': 3,'phf': 3,'spy': 3,'warezclient': 3,'warezmaster': 3,'sendmail': 3,'named': 3,'snmpgetattack': 3,'snmpguess': 3,'xlock': 3,'xsnoop': 3,'httptunnel': 3,
                           'buffer_overflow': 4,'loadmodule': 4,'perl': 4,'rootkit': 4,'ps': 4,'sqlattack': 4,'xterm': 4})


In [88]:
label_test_new =label_test.replace({ 'normal' : 0, 'neptune' : 1 ,'back': 1, 'land': 1, 'pod': 1, 'smurf': 1, 'teardrop': 1,'mailbomb': 1, 'apache2': 1, 'processtable': 1, 'udpstorm': 1, 'worm': 1,
                           'ipsweep' : 2,'nmap' : 2,'portsweep' : 2,'satan' : 2,'mscan' : 2,'saint' : 2
                           ,'ftp_write': 3,'guess_passwd': 3,'imap': 3,'multihop': 3,'phf': 3,'spy': 3,'warezclient': 3,'warezmaster': 3,'sendmail': 3,'named': 3,'snmpgetattack': 3,'snmpguess': 3,'xlock': 3,'xsnoop': 3,'httptunnel': 3,
                           'buffer_overflow': 4,'loadmodule': 4,'perl': 4,'rootkit': 4,'ps': 4,'sqlattack': 4,'xterm': 4})

In [90]:
label_train_new.value_counts()

0    67343
1    45927
2    11656
3      995
4       52
Name: class, dtype: int64

In [91]:
#now replace in the new values in dataset
train_new['class']=label_train_new
test_new['class']=label_test_new

In [92]:
print(train_new['class'].head())

0    0
1    0
2    1
3    0
4    0
Name: class, dtype: int64


In [93]:
print(test_new['class'].head())

0    1
1    1
2    0
3    2
4    2
Name: class, dtype: int64


#### Now split the dataset into 4 type of attacks. Each dataset will contain the respective attack data and the normal traffic

In [94]:
# create 4 dataset and drop the other attacks 
to_drop_DOS =[2,3,4]
to_drop_Probe =[1,3,4]
to_drop_R2L =[1,2,4]
to_drop_U2R =[1,2,3]

In [96]:
# train dataset
Probe_train = train_new[~train_new['class'].isin(to_drop_Probe)]
R2L_train=train_new[~train_new['class'].isin(to_drop_R2L)]
U2R_train=train_new[~train_new['class'].isin(to_drop_U2R)]

In [95]:
DOS_train = train_new[~train_new['class'].isin(to_drop_DOS)]

In [97]:
# test dataset
DOS_test = test_new[~test_new['class'].isin(to_drop_DOS)]

In [98]:
Probe_test = test_new[~test_new['class'].isin(to_drop_Probe)]

In [99]:
R2L_test=test_new[~test_new['class'].isin(to_drop_R2L)]
U2R_test=test_new[~test_new['class'].isin(to_drop_U2R)]

In [107]:
#print the shapes
print('Train:')
print('DOS :',DOS_train.shape)
print('Probe:', Probe_train.shape)
print('R2L:', R2L_train.shape)
print('U2R:', U2R_train.shape)

Train:
DOS : (113270, 123)
Probe: (78999, 123)
R2L: (68338, 123)
U2R: (67395, 123)


In [105]:
print('Test:')
print('DOS :',DOS_test.shape)
print('Probe:', Probe_test.shape)
print('R2L:', R2L_test.shape)
print('U2R:', U2R_test.shape)

Test:
DOS : (17171, 123)
Probe: (12132, 123)
R2L: (12596, 123)
U2R: (9778, 123)


# Step 2 : Feature Scaling

### Seperate the X and Y columns 

In [109]:
# create X and Y for train data 
X_Dos = DOS_train.drop('class',1)

In [111]:
Y_Dos =DOS_train['class']

In [112]:
X_Probe =Probe_train.drop('class',1)
Y_Probe=Probe_train['class']
X_R2L=R2L_train.drop('class',1)
Y_R2L=R2L_train['class']
X_U2R=U2R_train.drop('class',1)
Y_U2R=U2R_train['class']

In [113]:
# create X and Y for test data
X_Dos_test = DOS_test.drop('class',1)
Y_Dos_test =DOS_test['class']

In [114]:
X_Probe_test =Probe_test.drop('class',1)
Y_Probe_test=Probe_test['class']
X_R2L_test=R2L_test.drop('class',1)
Y_R2L_test=R2L_test['class']
X_U2R_test=U2R_test.drop('class',1)
Y_U2R_test=U2R_test['class']

In [115]:
#list of current features
columnNames= list(X_Dos)
colName_test=list(X_Dos_test)

## Normalization

In [116]:
from sklearn.preprocessing import StandardScaler

In [118]:
#train data
scale_Dos= StandardScaler().fit(X_Dos)

In [119]:
X_Dos = scale_Dos.transform(X_Dos)

In [179]:
X_Dos.shape

(113270, 122)

In [122]:
scale_Probe= StandardScaler().fit(X_Probe)
X_Probe = scale_Probe.transform(X_Probe)

In [123]:
scale_R2L= StandardScaler().fit(X_R2L)
X_R2L= scale_R2L.transform(X_R2L)

In [124]:
scale_U2R= StandardScaler().fit(X_U2R)
X_U2R= scale_U2R.transform(X_U2R)

In [125]:
# normalize test data
scale1 = StandardScaler().fit(X_Dos_test)
X_Dos_test = scale1 .transform(X_Dos_test)
scale2 = StandardScaler().fit(X_Probe_test)
X_Probe_test = scale2.transform(X_Probe_test)
scale3 = StandardScaler().fit(X_R2L_test)
X_R2L_test = scale3.transform(X_R2L_test)
scale4 = StandardScaler().fit(X_U2R_test)
X_U2R_test = scale4.transform(X_U2R_test)

#### check the standard deviation

In [126]:
X_Dos.std(axis=0)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 0., 1.,
       0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1.,
       1., 1., 1.])

Now our dataset is numerical and normalized. It is ready to use. This step is called Data Preprocesisng. Now we can use data mining methods to see which features are useful and remove the redundant features before building any machine learning model.

# Step 3: Feature Selection

We need only the features that are highly dependent on the response variable i.e the output label. We can choose any apropriate feature selection method. In this case, we have chosen Univariate Feature selection method using ANOVA F-Test and Recursive Feature Elimination (RFE) for Feature Ranking. ANOVA means Analysis of Variance.
It is important to find the 'Variance' of the Feature. The variance means how much the feature impacts the response variable. If the variance is low, it means the feature doesnt have any impact on the response and vice-versa.

In [128]:
from sklearn.feature_selection import SelectPercentile, f_classif

If the features are numerical, ANNOVA F-value method can be used. It determines if we group the numerical features by target vector then if the means of each group are significantly different.

SelectPercentile : selects features according to the percentile of the highest scores.

f_classif finds the ANNOVA F-value between the label/feature for classification tasks.

In [129]:
# set to ignore divide and invalid error
np.seterr(divide='ignore', invalid='ignore');

In [130]:
feature_selector = SelectPercentile(f_classif, percentile=10)

In [131]:
X_Dos_new = feature_selector.fit_transform(X_Dos,Y_Dos )


/Users/garima/anaconda3/lib/python3.7/site-packages/sklearn/feature_selection/univariate_selection.py:114: UserWarning: Features [ 16  44  63  66  68  86 114] are constant.
  UserWarning)


In [132]:
X_Dos_new.shape

(113270, 13)

#### 13 features were selected for DOS attack dataset by ANNOVA F-Test. We can list which features were selected by ANNOVA

In [133]:
feature= feature_selector.get_support()

In [134]:
newcolindex_dos= [i for i , x in enumerate(feature) if x]

In [135]:
newcolname_dos = list(columnNames[i] for i in newcolindex_dos)

In [136]:
print(newcolname_dos)

['logged_in', 'count', 'serror_rate', 'srv_serror_rate', 'same_srv_rate', 'dst_host_count', 'dst_host_srv_count', 'dst_host_same_srv_rate', 'dst_host_serror_rate', 'dst_host_srv_serror_rate', 'service_http', 'flag_S0', 'flag_SF']


In [137]:
# now find the features selected in other attacks datasets
X_Probe_new =feature_selector.fit_transform(X_Probe,Y_Probe )

/Users/garima/anaconda3/lib/python3.7/site-packages/sklearn/feature_selection/univariate_selection.py:114: UserWarning: Features [ 4 16] are constant.
  UserWarning)


In [138]:
X_Probe_new.shape

(78999, 13)

In [139]:
# Get the features selected for Probe
feature= feature_selector.get_support()
newcolindex_probe= [i for i , x in enumerate(feature) if x]
newcolname_probe = list(columnNames[i] for i in newcolindex_probe)
print(newcolname_probe)

['logged_in', 'rerror_rate', 'srv_rerror_rate', 'dst_host_srv_count', 'dst_host_diff_srv_rate', 'dst_host_same_src_port_rate', 'dst_host_srv_diff_host_rate', 'dst_host_rerror_rate', 'dst_host_srv_rerror_rate', 'protocol_type_icmp', 'service_eco_i', 'service_private', 'flag_SF']


In [140]:
# Get the features selected for R2L
X_R2L_new =feature_selector.fit_transform(X_R2L,Y_R2L )

/Users/garima/anaconda3/lib/python3.7/site-packages/sklearn/feature_selection/univariate_selection.py:114: UserWarning: Features [  4  16  43  44  46  47  48  49  50  51  54  57  58  62  63  64  66  67
  68  70  71  72  73  74  76  77  78  79  80  81  82  83  86  87  89  92
  93  96  98  99 100 107 108 109 110 114] are constant.
  UserWarning)


In [141]:
X_R2L_new.shape

(68338, 13)

In [142]:
feature= feature_selector.get_support()
newcolindex_r2l= [i for i , x in enumerate(feature) if x]
newcolname_r2l = list(columnNames[i] for i in newcolindex_r2l)
print(newcolname_r2l)

['src_bytes', 'dst_bytes', 'hot', 'num_failed_logins', 'is_guest_login', 'dst_host_srv_count', 'dst_host_same_src_port_rate', 'dst_host_srv_diff_host_rate', 'service_ftp', 'service_ftp_data', 'service_http', 'service_imap4', 'flag_RSTO']


In [143]:
# Get the features selected for U2R
X_U2R_new =feature_selector.fit_transform(X_U2R,Y_U2R )

/Users/garima/anaconda3/lib/python3.7/site-packages/sklearn/feature_selection/univariate_selection.py:114: UserWarning: Features [  4  16  43  44  46  47  48  49  50  51  54  57  58  62  63  64  66  67
  68  70  71  72  73  74  75  76  77  78  79  80  81  82  83  86  87  89
  92  93  96  98  99 100 107 108 109 110 114] are constant.
  UserWarning)


In [144]:
X_U2R_new.shape

(67395, 13)

In [145]:
feature= feature_selector.get_support()
newcolindex_u2r= [i for i , x in enumerate(feature) if x]
newcolname_u2r = list(columnNames[i] for i in newcolindex_u2r)
print(newcolname_u2r)

['urgent', 'hot', 'root_shell', 'num_file_creations', 'num_shells', 'srv_diff_host_rate', 'dst_host_count', 'dst_host_srv_count', 'dst_host_same_src_port_rate', 'dst_host_srv_diff_host_rate', 'service_ftp_data', 'service_http', 'service_telnet']


In [159]:
# map the features for test sets also
test_probe_new = X_Probe_test[:,newcolindex_probe]
test_r2l_new = X_R2L_test[:,newcolindex_r2l]
test_u2r_new = X_U2R_test[:,newcolindex_u2r]

In [ ]:
test_dos_new = X_Dos_test[:,newcolindex_dos]

In [158]:
test_dos_new.shape

(17171, 13)

In [160]:
test_u2r_new.shape

(9778, 13)

# Step 4: Build Model

## Build Deep Learning Fully connected Neural Network

We will now build ML models with the new feature set, for all the four datasets. We can also build another model with all the features and compare both the models.

In [146]:
from keras.models import Sequential

Using TensorFlow backend.


In [148]:
from keras.layers import Dense

In [165]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score

In [149]:
#lets first build NN model for DOS attack

In [150]:
# Define the NN architecture for Dos attack dataset
model_dos= Sequential()
model_dos.add(Dense(24, input_dim=13, activation='relu'))
model_dos.add(Dense(12,activation='relu'))
model_dos.add(Dense(8, activation='relu'))
model_dos.add(Dense(1, activation='sigmoid'))

Instructions for updating:
If using Keras pass *_constraint arguments to layers.


In [151]:
# compile the model
model_dos.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'] )

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


In [152]:
model_dos.fit(X_Dos_new, Y_Dos, epochs=50, batch_size=100)


Epoch 1/50
113270/113270 [==============================] - 1s 10us/step - loss: 0.1188 - accuracy: 0.9686
Epoch 2/50
113270/113270 [==============================] - 1s 8us/step - loss: 0.0693 - accuracy: 0.9778
Epoch 3/50
113270/113270 [==============================] - 1s 8us/step - loss: 0.0630 - accuracy: 0.9787
Epoch 4/50
113270/113270 [==============================] - 1s 8us/step - loss: 0.0597 - accuracy: 0.9793
Epoch 5/50
113270/113270 [==============================] - 1s 8us/step - loss: 0.0576 - accuracy: 0.9797
Epoch 6/50
113270/113270 [==============================] - 1s 8us/step - loss: 0.0562 - accuracy: 0.9799
Epoch 7/50
113270/113270 [==============================] - 1s 8us/step - loss: 0.0548 - accuracy: 0.9804
Epoch 8/50
113270/113270 [==============================] - 1s 8us/step - loss: 0.0537 - accuracy: 0.9808
Epoch 9/50
113270/113270 [==============================] - 1s 8us/step - loss: 0.0529 - accuracy: 0.9814
Epoch 10/50
113270/113270 [=================

### Prediction Accuracy and Evaluation

In [153]:
# training accuracy
train_acc = model_dos.evaluate(X_Dos_new, Y_Dos, verbose=0)

In [156]:
print("%s: %.2f%%" % (model_dos.metrics_names[1], train_acc[1]*100))

accuracy: 98.57%


In [162]:
# Prediction
predict = model_dos.predict_classes(test_dos_new)

In [166]:
test_acc = accuracy_score(Y_Dos_test, predict)

In [167]:
print('Prediction Accuracy: %f' %test_acc)

Prediction Accuracy: 0.808864


### Cross Validation: Accuracy, Precision, Recall and F-Measure

#### Lets us print the NN model metrics and find the Precision, Recall and F-measure.

## accuracy: (tp + tn) / (p + n)
## precision tp / (tp + fp)
## recall: tp / (tp + fn)
## f1: 2 tp / (2 tp + fp + fn)

In [168]:
#Precision
precision = precision_score(Y_Dos_test, predict)
print('Precision: %f' %precision)

Precision: 0.839896


In [169]:
# Recall
recall = recall_score(Y_Dos_test, predict)
print('Recall: %f' %recall)

Recall: 0.691957


In [170]:
# F1 Score
f1= f1_score(Y_Dos_test, predict)
print('f1 Score: %f' %f1)

f1 Score: 0.758783


In [171]:
# ROC AUC
roc= roc_auc_score(Y_Dos_test, predict)
print('ROC-AUC Score: %f' %roc)

ROC-AUC Score: 0.795314


In [173]:
from sklearn.metrics import confusion_matrix

In [174]:
# Confusion matrix
matrix = confusion_matrix(Y_Dos_test, predict)
print(matrix)

[[8727  984]
 [2298 5162]]


In [175]:
# Now we can similarly build NN model for the rest of three datasets, 
#Each dataset has 13 features so architecture will remain same

In [ ]:
# We can also build NN model with all the features and compare the results

In [180]:
# Define the NN architecture for DOS attack dataset with all the 122 features.
# accordingly define the number of nodes in hidden layer
model_dos_all = Sequential()
model_dos_all.add(Dense(256, input_dim=122, activation='relu'))
model_dos_all.add(Dense(128,activation='relu'))
model_dos_all.add(Dense(64, activation='relu'))
model_dos_all.add(Dense(1, activation='sigmoid'))

In [181]:
model_dos_all.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [182]:
model_dos_all.fit(X_Dos, Y_Dos, epochs=50, batch_size=100)

Epoch 1/50
113270/113270 [==============================] - 2s 19us/step - loss: 0.0162 - accuracy: 0.9961
Epoch 2/50
113270/113270 [==============================] - 2s 16us/step - loss: 0.0042 - accuracy: 0.9987
Epoch 3/50
113270/113270 [==============================] - 2s 16us/step - loss: 0.0030 - accuracy: 0.9989
Epoch 4/50
113270/113270 [==============================] - 2s 16us/step - loss: 0.0024 - accuracy: 0.9991
Epoch 5/50
113270/113270 [==============================] - 2s 16us/step - loss: 0.0024 - accuracy: 0.9992
Epoch 6/50
113270/113270 [==============================] - 2s 16us/step - loss: 0.0020 - accuracy: 0.9993
Epoch 7/50
113270/113270 [==============================] - 2s 16us/step - loss: 0.0022 - accuracy: 0.9994
Epoch 8/50
113270/113270 [==============================] - 2s 16us/step - loss: 0.0016 - accuracy: 0.9994
Epoch 9/50
113270/113270 [==============================] - 2s 16us/step - loss: 0.0015 - accuracy: 0.9993
Epoch 10/50
113270/113270 [==========

In [183]:
# training accuracy
train_acc = model_dos_all.evaluate(X_Dos, Y_Dos, verbose=0)

In [184]:
print("%s: %.2f%%" % (model_dos.metrics_names[1], train_acc[1]*100))

accuracy: 99.97%


In [185]:
# Check the Test Accuracy
# Prediction
predict = model_dos_all.predict_classes(X_Dos_test)

In [186]:
print('Prediction Accuracy: %f' %test_acc)

Prediction Accuracy: 0.808864


In [187]:
# The training accuracy is higher than previous model but test accuracy is low, so this may be due to overfitting

In [188]:
#Precision
precision = precision_score(Y_Dos_test, predict)
print('Precision: %f' %precision)

Precision: 0.987195


In [189]:
# Recall
recall = recall_score(Y_Dos_test, predict)
print('Recall: %f' %recall)

Recall: 0.785389


In [190]:
# F1 Score
f1= f1_score(Y_Dos_test, predict)
print('f1 Score: %f' %f1)

f1 Score: 0.874804


In [192]:
# ROC AUC
roc= roc_auc_score(Y_Dos_test, predict)
print('ROC-AUC Score: %f' %roc)

ROC-AUC Score: 0.888781


In [193]:
#Recall and ROC-AUC was better in this case than previous model

In [194]:
# Confusion matrix
matrix = confusion_matrix(Y_Dos_test, predict)
print(matrix)

[[9635   76]
 [1601 5859]]
